cpp/src/strings/json/json_path.cu

```diff
@@ -231,17 +231,30 @@ class parser {
 /**
  * @brief Output buffer object.  Used during the preprocess/size-computation step
  * and the actual output step.
+ *
+ * There is an important distinction between two cases:
+ *
+ * - producing no output at all. that is, the query matched nothing in the input.
+ * - producing empty output. the query matched something in the input, but the
+ *   value of the result is an empty string.
+ *
+ * The `has_output` field is the flag which indicates whether or not the output
+ * from the query should be considered empty or null.
+ *
  */
 struct json_output {
   size_t output_max_len;
   size_t output_len;
-  int element_count;
+  bool has_output;
   char* output;
 
   constexpr void add_output(const char* str, size_t len)
   {
     if (output != nullptr) { memcpy(output + output_len, str, len); }
     output_len += len;
+    // set this to true even if the string is empty. it implies we have
+    // seen actual output
+    has_output = true;
```

Comment:
Any kind of bool like this is an immediate hint to use an `optional<T>` instead.